stix_shifter_modules/crowdstrike_logscale/stix_transmission/connector.py

```diff
@@ -0,0 +1,150 @@
+import re
+import json
+from flatten_json import unflatten
+from stix_shifter_utils.modules.base.stix_transmission.base_json_sync_connector import BaseJsonSyncConnector
+from stix_shifter_utils.utils.error_response import ErrorResponder
+from stix_shifter_utils.utils import logger
+from .api_client import APIClient
+
+
+# LogScale event fields starts with @timestamp and @error_msg will be overwritten while un-flattening.
+# These fields are excluded while un flattening
+DS_FLATTEN_KEY_EXCLUDE = ["@timestamp", "@error_msg"]
+
+
+class Connector(BaseJsonSyncConnector):
+
+    def __init__(self, connection, configuration):
+        self.api_client = APIClient(connection, configuration)
+        self.logger = logger.set_logger(__name__)
+        self.connector = __name__.split('.')[1]
+
+    async def ping_connection(self):
+        """
+        Ping the endpoint
+        :return: return_object, dict
+        """
+        return_obj = {}
+        response_dict = {}
+        try:
+            response = await self.api_client.ping_data_source()
+            response_code = response.code
+            response_str = response.read().decode('utf-8')
+            if response_code == 200:
+                return_obj['success'] = True
+            else:
+                return_obj = self.exception_response(response_code, response_str)
+
+        except Exception as ex:
+            if "timeout_error" in str(ex):
+                response_dict['code'] = 408
+            response_dict['message'] = str(ex)
+            self.logger.error('error while pinging: %s', ex)
+            ErrorResponder.fill_error(return_obj, response_dict, ['message'], connector=self.connector)
+        return return_obj
+
+    async def create_results_connection(self, query, offset, length, metadata=None):
+        """
+        Fetching the results using query, offset and length
+        :param query: str, Data Source query
+        :param offset: str, Offset value
+        :param length: str, Length value
+        :param metadata: dict
+        :return: return_obj, dict
+        """
+        return_obj = {}
+        data = []
+        offset = int(offset)
+        length = int(length)
+        try:
+            if isinstance(query, str):
+                query = json.loads(query)
+            source = query.get('source')
+
+            response_wrapper = await self.api_client.get_search_results(json.dumps(query))
+            response = response_wrapper.read().decode('utf-8')
+            if response_wrapper.code == 200:
+                return_obj['success'] = True
+                data += json.loads(response)
+            else:
+                return_obj = self.exception_response(response_wrapper.code, response)
+                data = []
+
+            if data:
+                # LogScale response is not in sorted order, sorting events based on @timestamp
```

Comment:
its not clear why the sorting is needed based on timestamp. Is it for pagination? Do you need the event sorted to get next set of results? If yes, how do you save the last timestamp for the subsequent query?

We basically need to clarify on how to paginate the results once the length is reached.

Reply:
This connector doesn't have pagination. The results which are returned from API are unordered for the provided timestamp.  Since this is synchronous connector, the entire results will be fetched for each run in CP4S. In order to apply offset and length for each run in CP4S, data is sorted.